[수한] 6주차 문제 풀이

sem1308 opened the pull request:
알고리즘 풀이 제출합니다

Commit:
SW 2115 벌꿀채취

- 109 ms
- 구현, 조합

=== 로직 ===
1. x,y 에서 오른쪽으로 M만큼 벌통을 놓았을 때 가능한 최대 채취 양 저장
2. 0,0 에서 시작하여 두 일꾼이 수확할 수 있는 최대 채취 양 구함

Commit:
SW 2112 보호필름

- 225 ms, 229ms
- 구현, 부분집합, 백트래킹

=== 로직 ===
1. 각 막 마다 약품을 투여 안하는 경우, 약품 A를 투여하는 경우, 약품 B를 투여하는 경우 총 3가지 경우 완전 탐색 (부분집합)
2. 모든 막에 대한 처리 후 세로 방향마다 특성이 K개 연속되는지 체크
3. 검사를 통과하면 답 update
4. 재귀를 돌며 현재 약품 투여 수가 답보다 같거나 크다면 return (백트래킹)

Commit:
SW 2383 점심 식사시간

- 113 ms
- 구현, 부분집합, Queue

=== 로직 ===
1. 각 사람마다 2개의 계단 중 1개로 간다고 생각
2. 각 계단마다 사람 배치
3. 해당 계단에 도착하는 사람들의 거리 저장 후 정렬
4. 시간을 0부터 1씩 증가해가며 시뮬레이션
5. 현재 계단에 있는 사람들 처리 - 현재 계단에 있던 사람들의 다 내려갔을 때의 시간이 현재 시간이면 pop
   - pop할 때 계단 내려간 사람 수 증가
6. 기다리던 사람들 계단에 추가 (계단은 Queue로 구현) (다 내려갔을 때의 시간을 삽입)
7. 계단 내려간 사람 수가 그 계단에 배치된 사람 수와 같다면 그 다음 계단 처리
8. 모든 계단 처리 후 각 계단 처리 시간의 최댓값이 현재 최소 시간보다 짧다면 update

Commit:
SW 2477 차량정비소

- 143 ms
- 구현, 시뮬레이션, 자료구조(Queue)

=== 로직 ===
(고객 클래스 생성 - 고객 번호, 고객이 방문한 접수 창구 번호, 고객 처리가
끝나는 시간)
(접수, 정비 창구는 queue로 구현)
1. 시간을 0부터 시작해서 1씩 증가시키며 시뮬레이션
2. 현재 시간이 고객 도착시간과 같으면 접수 기다리는 리스트에 추가
3. 접수창구 -> 정비창구 순서로 처리
   3.1. 접수창구
    - 낮은 번호의 접수 창구부터 루프 (창구번호 작은 곳 우선)
    - 지금 시간이 접수 끝난 시간인 경우
        - 정비 창구 기다리는 곳에 고객 추가
        - 해당 접수 창구 비우기
    - 접수 창구 빈 자리에 기다리는 사람 넣기
   3.2. 정비창구
    - 낮은 번호의 정비 창구부터 루프 (창구번호 작은 곳 우선)
    - 지금 시간이 정비 끝난 시간인 경우
        - 접수창구, 정비창구에 들른 고객 번호 합 갱신
        - 처리가 모두 끝난 고객 수 증가
        - 해당 정비 창구 비우기
    - 정비 창구 빈 자리에 기다리는 사람 넣기
4. 찾으려는 접수 창구, 정비 창수에 왔었던 고객 번호 합 출력

Commit:
SW 5656 벽돌깨기

- 729 ms, 1409 ms
- 구현, 시뮬레이션

=== 로직 ===
1. 벽돌 떨어뜨릴 위치 구함
2. 시뮬레이션

배열 : 729 ms
연결리스트 : 1409ms

gangintheremark reviewed (approved):
구체적인 변수명이 인상깊네요 코드가 전체적으로 깔끔하니 보기 좋습니다 훠훠 수고하셨습니다

Commit:
SW 5656 벽돌깨기

- 187 ms
- 구현, 시뮬레이션

=== 로직 ===
1. 벽돌 떨어뜨릴 위치 구함
2. 위치 선정 후 바로 시뮬레이션
3. 시뮬레이션을 통해 깨진 벽돌 개수구함
